### ISSUE TITLE
job-manager: scheduler interface

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1697]
The `job-manager` module needs to call out to the scheduler to allocate resources, as discussed in #1666.   Cancellation of scheduler `allocate` requests was also brought up in #688.  Let's try to capture the details of the `job-manager` - `sched` interface here.  Here's an initial proposal based on my recollection of our discussions for a start:

**sched.allocate**
*request*
* jobid
* submitting userid
* submit time
* priority
* reference to jobspec in the KVS

*response*
* jobid
* reference to `R` in the KVS (placed there by sched)

`allocate` requests are sent to sched in order of 1) highest priority job, then 2) lowest numbered job ID (with FLUIDs this is roughly time ordered)

A maximum number of outstanding `allocate` requests are allowed.  This number is regulated by sched through a credit-based scheme (see below).

**sched.free**

*request*
* jobid
* reference to `R` in the KVS

*response*
* empty (success/failure)

**sched.cancel**

*request*
* jobid

There is no direct response.  If there is an outstanding `allocate` request for this job, it receives an error response.

**sched.get-credits**

Request:
* number of credits requested
* priority level

Response
* number of credits granted (1 or more)

A job manager (or other sched client) may only send one `allocate` request per credit, then must request more credits.  The scheduler may throttle the flow of `allocate` requests and limit concurrency by delaying the response to a request for credits.   It must implement a per-client count of credits issued, and return an error to `allocate` requests when there are no credits.

To allow high priority jobs to be accepted when the scheduler has already issued all its credits, a request for credits may contain a priority.  If credits are not available, the scheduler may cancel lower priority requests to allow the higher priority requests to be accepted.   Requests for credits are handled first in priority order, then in FCFS order.

To avoid deadlock, `free` requests must not be subject to credits.

**Comment by SteVwonder** [comment]
Generally LGTM.  Couple of things that popped into my head while reading.

> sched.cancel: There is no direct response. If there is an outstanding allocate request for this job, it receives an error response.

Are RPCs guaranteed to be non-overtaking?  If not, then a job manager module could send (sched.allocate, sched.cancel, sched.allocate) and they could be received out-of-order (sched.allocate, sched.allocate, sched.cancel).  Resulting in the job manager thinking it has an outstanding allocate request and the scheduler thinking the allocate request was cancelled.

> If credits are not available, the scheduler may cancel lower priority requests to allow the higher priority requests to be accepted.

How does a scheduler cancel lower priority requests for credits?  Does it respond with an error?  Or maybe responds with 0 credits.

Can job managers issue multiple requests for credits?  This would be potentially useful for the scenario where multiple priorities are in the job manager's queue.  The job manager could issue one credit request for each priority category.  If a job manager can only have one outstanding request for credits at a time, I think this will result in some blocking of high priority jobs.  For example, if a job manager has an outstanding, low priority request for credits and then receives a new high priority job, must the job manager wait for a response from the scheduler to its original, low-priority credit request before it can issue a new high-priority request (potentially for only a single credit)?

**Comment by dongahn** [comment]
Great start. I few questions/comments.

> 2) lowest numbered job ID (with FLUIDs this is roughly time ordered)

I think the roughly time-ordered property can lead to a side effect. If the workload assumes strict first come first served order (like in the case of the Cancer Moonshot), this can lead to incorrect out of order execution. Of course, the scheduler itself can reorder jobs in its own queue based on the submit-time information, but if the earlier job is already allocated before the later job allocation is requested, the reordering won't work.

Is there a FLUID property that job manager can latch onto so that it can actually send allocate requests in the strict submit-time order? I understand why you want to do this in FLUIDs order for efficiency reason, though.

> sched.free

This is not really an interface question but there will be a guarantee that for each allocated R, will there be a set of frees whose aggregate Rs match with the allocated R? Having this guarantee will make the scheduler logic simple by not introducing a garbage collection of some sort. 

> sched.cancel

Just make sure I understand its semantics, it should only be meaningful before the jobid is allocated, correct? 

In any case, I think  a race condition can arise here, though. The corresponding allocate request has already gotten an R (R has already been dumped to KVS) but haven't yet received a response. So, for the case like this, I have to guess it would be better to have an error reply for the cancel request? 


Finally, the direction I will want to pursue with the resource matching service is to allow multi-requests to reduce the individual handler invocation overhead. Introducing things like match-allocate-multi. This might ease the head of line blocking at the messaging level. Did you consider adding things like that for HTC workloads?

**Comment by dongahn** [comment]
> Are RPCs guaranteed to be non-overtaking? 

This is really a good question. Typically, when you talk about a pair-wise communication, allowing overtaking makes parallel and distributed programming much harder to reason about.

**Comment by dongahn** [comment]
> Request:
- number of credits requested
- priority level
Response
- number of credits granted (1 or more)

So, this is 1 to the number of credits requests? Would you allow the response to send more than requested credits -- if the scheduler doesn't want throttling at all?

**Comment by SteVwonder** [comment]
> will there be a set of frees whose aggregate Rs match with the allocated R. Having this guarantee will make the scheduler logic simple by not introducing a garbage collection of some sort.

I agree.  Some entity within Flux should be responsible for declaring that a job has had all of its resources completely free'd (IIRC the job shell is the one currently slated to do this), and then the scheduler should be notified that it is safe to now completely forget about a given job.  In the current diagram on the whiteboard, we have arrows going between the job manager and scheduler labelled `free`, `allocate`, and `job_done` (although that name could be improved).

**Comment by garlick** [comment]
> Is there a FLUID property that job manager can latch onto so that it can actually send allocate requests in the strict submit-time order?

They are in order as long as they are submitted on the same rank, so a workflow manager acting as the initial program should be able to get that.

> This is not really an interface question but there will be a guarantee that for each allocated R, will there be a set of frees whose aggregate Rs match with the allocated R?

Let's go with "yes" on that one. :-)

> [on `sched.cancel`] Just make sure I understand its semantics, it should only be meaningful before the jobid is allocated, correct?

Right.  This would cause the allocate request to get an error response.   You are right, the allocate request could get an `R` before it can be cancelled, in which case we (job manager) probably know the job is being cancelled and can `free` the `R`.  As proposed, there is no way to give an error response to the `cancel` directly but I don't think it's necessary.

**Comment by garlick** [comment]
> Finally, the direction I will want to pursue with the resource matching service is to allow multi-requests to reduce the individual handler invocation overhead. Introducing things like match-allocate-multi. This might ease the head of line blocking at the messaging level. Did you consider adding things like that for HTC workloads?

I think batching multiple requests in one message is functionally equivalent to sending multiple request messages back to back, as long as a you dequeue all of them right away.  Either way I think it means you have to do something with multiple requests that arrive simultaneously (queue some and process one?  start all of them asynchronously?)   Bottom line, if you use the socket as a work queue, but you have work mixed with other message types, you may have head of line blocking.  Best to dequeue from the socket and manage a work queue separately.

**Comment by garlick** [comment]
> * number of credits granted (1 or more)
>
> So, this is 1 to the number of credits requests? Would you allow the response to send more than requested credits -- if the scheduler doesn't want throttling at all?

I think responses should be from 1 to the number requested.  Always granting the number requested would be no throttling or concurrency limiting.  The main point I wanted to make was that if there are no credits, the response should wait until there is at least one. (Don't return zero or error in that case, or the other end will have to resort to polling).

**Comment by dongahn** [comment]
> I think batching multiple requests in one message is functionally equivalent to sending multiple request messages back to back, as long as a you dequeue all of them right away. Either way I think it means you have to do something with multiple requests that arrive simultaneously (queue some and process one? start all of them asynchronously?) Bottom line, if you use the socket as a work queue, but you have work mixed with other message types, you may have head of line blocking. Best to dequeue from the socket and manage a work queue separately.

Yes, it is functionally equivalent. But since there would be a limit on Flux's message injection rate, thought is, an one bulk allocate request would be much more efficient than having lots of individual requests. Wondering what throughputs we will see with `sleep 0` jobs, and how much portion of that would be due to messaging.

**Comment by garlick** [comment]
Ah I see - well, that sort of optimization feels premature in this case.  I guess we can always add that sort of thing later if it becomes a problem!

**Comment by dongahn** [comment]
Yeah, I wasn't suggesting we should add this as the initial proposal. Just commented as our future reference.

**Comment by SteVwonder** [comment]
So this may be more trouble than it is worth (and it completely flips the scheme to be scheduler-driven rather than job manager-driven), but I was thinking through the idea of adding the ability for the scheduler to "query" the job manager for pending jobs (i.e., not held by the user or blocked on a dependency) matching a particular "description" rather than accepting allocate requests from the job manager in order of job priority.

At today's meeting, we said that two instructive examples are fair-share priority with backfilling and greedy scheduling (for high-throughput workloads).  So I'll use those as examples to illustrate the idea.

For the greedy case, assuming there are 5 nodes currently free and a queue depth of 100, I imagine the scheduler submitting a query like (in SQL-ish format for convenience):
```SQL
SELECT TOP 100
WHERE num_nodes <= 5
ORDER BY primary_priority, num_nodes
```

For the backfilling case, assuming the reservation depth is 25, I imagine the scheduler submitting a query like:
```SQL
SELECT TOP 25
ORDER BY primary_priority, secondary_priority
```
and then attempting to schedule all of those jobs.  If any of them are successfully scheduled at the current time (i.e., not a reservation), then another query of the same form will need to be submitted to ensure that exactly 25 reservations are made.  Then to get jobs that are valid for backfilling (and assuming the queue depth is 100, the last job to get a reservation had a primary_priority of 16 and secondary_priority of 1000, and 20 nodes are currently available) then I imagine the scheduler issuing a query of the form:
```SQL
SELECT TOP 75 -- 100 queue depth - 25 reservations
WHERE (primary_priority <= 16) and (secondary_priority <= 1000) and (num_nodes <= 20)
ORDER BY primary_priority, secondary_priority
```

Pros:

- Easily supports schedulers/policies like greedy that wants sorting & filtering based on properties other than strictly priority
- Enables bulk requests for jobs
- If we can provide a nice query interface that is usable by multiple schedulers, then we get two benefits:
  - We push the complexity of filtering/sorting/select jobs out of the scheduler into a reusable module.
  - We can completely change the back-end implementation without rewriting the scheduler.  So we could start with a simple implementation that assumes the scheduler and job manager are both serialized and running in the same process, and then evolve the implementation to support parallelizing the job manager across nodes/ranks.

Cons:

- An implementation with a distributed job manager will require coordination to enforce the "LIMIT" and "ORDER BY" clauses.
- In the case of any follow-up queries, the scheduler will need to check the job IDs to ensure that a job wasn't returned for both the original and follow-up query.

**Comment by SteVwonder** [comment]
After a discussion over coffee with @garlick and @grondo, I think we settled on a nice interface that handles the greedy and fair-share use-cases.

So first off, we came to the conclusion that the above SQL-esque query functionality would be difficult/messy to implement due to moldable/malleable jobs.  If your moldable jobs' walltime is a function of the number of nodes allocated to them, then filtering based on the block(s) of free resources that you have available (e.g., 20 nodes for 1 hour OR 10 nodes for 2 hours OR 1 node for 24 hours) becomes way more complex.  It also would require the responding entity to parse and handle jobspec.

Instead, we think that allowing the scheduler to perform another level of ordering/sorting (on top of the two priorities provided by the job manager) makes sense.  For example, in the "greedy" case, the scheduler could request an infinite number of jobs from the job manager, and then sort the jobs based on number of nodes required.  The scheduler could then reply to the job manager module's `allocate` request with a JSON payload containing this new ordering plus other metadata like estimate start time, confidence in the predicted start time (e.g., guaranteed vs 80% confidence), and reason for not scheduling yet (e.g., insufficient licenses or nodes).  This reply would be a part of a multi-response RPC, so the `allocate` request would remain open and the scheduler could send a new response every-time the metadata changes.

In the case of a fair-share scheduler, it would only request a `queue-depth` number of jobs from the job manager as well as keep the same order/priority as provided by the job manager.  Even though it is not changing the order, it would still respond with metadata since it can provide an estimated start time based on job reservations and a reason for not scheduling yet.

**Comment by SteVwonder** [comment]
> Why can't the priority plugin (or module) to the job manager enable the greedy case, though?

That's a great question.  I thought we came up with a scenario where that wouldn't work, but it is escaping me currently. We spent a lot of time discussing the scheduler-driven query idea, so maybe by the time we looped back around to the original job manager-driven interface, we missed the obvious solution.

Originally, we were trying to avoid the scheduler requiring an infinite queue depth. I think we concluded that was a necessary evil in the greedy case, since if you limit the queue depth, you might miss out on jobs that are small enough to backfill.  Since an infinite queue is required either way, we wouldn't gain anything from adding an additional re-order on the scheduler-side.  I guess that's ultimately a good thing though, since having to track a scheduler queue order on top of two priorities wasn't looking too appealling.  (I do still like @garlick's suggestion of having the scheduler communicate metadata back to the job manager via a multi-response RPC).

> Are you worry about a case where both fair share priority and greedy sorting must be mixed? Do we anticipate that this would be a real use case?

Probably not, but if it does end up being a use case, I wonder if the greedy piece could be added as another weighted component in the fair-share linear function.

**Comment by grondo** [comment]
> Isn't the whole purpose of such an external priority function is to order jobs anyway needed for the target scheduler? 

I think we identified that for cases where scheduler priority isn't a function of submit order at all, trying to use an off-scheduler priority plugin would not be a good pattern. This is because the priority needs to be recalculated on every system state change. So for the greedy case, when a job is scheduled or completes, the scheduler would have to cancel all outstanding `alloc` requests, somehow update the priority plugin with the next "best" job size, and trigger a rerun of the priority plugin in the job manager which would issue new series of `alloc` requests up to the queue depth.

@garlick and @SteVwonder also identified that we should probably design the system so that high throughput scheduling is optimized, and other, more complex algorithms are possible. In fair-share for example, the pre-calculated priorities of the job-manager may be good enough to be one of the factors in the fair-share calculation, and the full fair share priority can be updated by an asynchronous background process, so we may not need a priority plugin with advanced knowledge of scheduler internals for that case either. (Not to say that priority plugin to the job-manager would not be supported, though)

**Comment by dongahn** [comment]
> I think we identified that for cases where scheduler priority isn't a function of submit order at all, trying to use an off-scheduler priority plugin would not be a good pattern. This is because the priority needs to be recalculated on every system state change. So for the greedy case, when a job is scheduled or completes, the scheduler would have to cancel all outstanding alloc requests, somehow update the priority plugin with the next "best" job size, and trigger a rerun of the priority plugin in the job manager which would issue new series of alloc requests up to the queue depth.

It seems what’s being proposed is to add a sort-of bypass logic and allow the scheduler to receive all of the jobs from the job manager -- essentially queue-depth=infinite, then? In this case, no credit-based throttling will be imposed? How would the RPC subsystem perform when you have potentially 10s of 1000s of alloc requests being issues and canceled on every system state change?

> @garlick<https://github.com/garlick> and @SteVwonder<https://github.com/SteVwonder> also identified that we should probably design the system so that high throughput scheduling is optimized, and other, more complex algorithms are possible. In fair-share for example, the pre-calculated priorities of the job-manager may be good enough to be one of the factors in the fair-share calculation, and the full fair share priority can be updated by an asynchronous background process, so we may not need a priority plugin with advanced knowledge of scheduler internals for that case either. (Not to say that priority plugin to the job-manager would not be supported, though)

This sounds good. The only thing is we need to pencil in the “interface” between the fair share priority module and the job manager. As far as that’s defined (as part of KVS job schema?), I can see we can even use other fair share priority code (e.g., slurm) to effect the fair share policy on job manager’s job queue.

**Comment by SteVwonder** [comment]
> This is because the priority needs to be recalculated on every system state change. So for the greedy case, when a job is scheduled or completes, the scheduler would have to cancel all outstanding alloc requests, somehow update the priority plugin with the next "best" job size, and trigger a rerun of the priority plugin in the job manager which would issue new series of alloc requests up to the queue depth.

Right, if you wanted to run greedy optimally with a limited queue depth, then this would have to be the process.

> It seems what’s being proposed is to add a sort-of bypass logic and allow the scheduler to receive all of the jobs from the job manager -- essentially queue-depth=infinite, then? In this case, no credit-based throttling will be imposed? How would the RPC subsystem perform when you have potentially 10s of 1000s of alloc requests being issues and canceled on every system state change?

I was thinking that a scheduler that holds a queue with all the jobs and a priority function based on the absolute size of the job (not the "best" size) would avoid a storm of cancel/alloc requests, but in exchange, the scheduler would have to handle the complexity that comes with so many jobs (maybe more advanced data structures could help here).  The scheduler would also require a pending/outstanding RPC for each job, which could be an issue.

**Comment by garlick** [comment]
Maybe greedy could have a tunable high and low water mark for queue depth, and just issue credits up to the high water mark whenever utilization falls below a configured threshold, then shrink back down?

Another idea is maybe it would be useful for sched to able to respond to an alloc request with "hold this job with tag X" and then have a sched to job-manager request to "release jobs with tag X"?   Just a way to avoid having temporarily unrunnable, but high priority jobs constantly flung at the scheduler?

**Comment by grondo** [comment]
This issue can be closed I think, as there now exists a job-manager/scheduler interface. We can open new issues for changes or enhancements to the existing protocol.
